enclave/src/bpf/enclave.bpf.c

```diff
@@ -0,0 +1,299 @@
+	u32 runtime_key = hash(comm, TASK_COMM_LEN);
+	u32 *runtime_lookup = bpf_map_lookup_elem(&runtimes, &runtime_key);
+
+	if (runtime_lookup) {
+		bpf_printk("found runtime init process: %d\n", pid);
+		err = add_container(child);
+		if (err < 0)
+			return err;
+		return 0;
+	}
+
+	u64 process_key = task_pid(parent);
+	struct process *parent_lookup = bpf_map_lookup_elem(&processes, &process_key);
+	if (!parent_lookup)
+		/* Nothing to do */
+		return 0;
+
+	bpf_printk("found parent containerized process: %d\n", process_key);
+
+	u32 container_key = parent_lookup->container_id;
+	u32 *container_lookup = bpf_map_lookup_elem(&containers, &container_key);
+	if (!container_lookup) {
+		/* Shouldn't happen */
+		return -EPERM;
+	}
+
+	struct process new_p = {
+		.container_id = container_key
+	};
+
+	bpf_printk("adding containerized process: %d\n", pid);
+	err = bpf_map_update_elem(&processes, &pid, &new_p, 0);
+	if (err < 0)
+		return err;
+
+	return 0;
+}
+
+/*
+ * get_policy_level - find the policy level for the given process.
+ * @pid: the PID of the process to find the policy for
+ *
+ * Return: corresponding policy level (or POLICY_LEVEL_NOT_FOUND when the
+ * process is not containerized, or POLICY_LEVEL_LOOKUP_ERROR when the state
+ * of BPF maps is inconsistent).
+ *
+ * TODO: Think of some better way to handle the POLICY_LEVEL_LOOKUP_ERROR - if
+ * that value is ever returned, it means that the container/process
+ * registration went wrong and we have insonsistent data.
+ */
+static __always_inline enum container_policy_level get_policy_level(u64 pid)
+{
+	int err;
+
+	struct process *p = bpf_map_lookup_elem(&processes, &pid);
+	if (!p)
+		return POLICY_LEVEL_NOT_FOUND;
+
+	struct container *c = bpf_map_lookup_elem(&containers, &p->container_id);
+	if (!c)
+		/* Shouldn't happen */
+		return POLICY_LEVEL_LOOKUP_ERR;
+
+	return c->policy_level;
+}
+
+/*
+ * BPF programs
+ * ============
+ */
+
+/*
+ * NOTE(mrostecki): Apparently, to monitor **all** the processes in the system,
+ * I had to use both the `sched_process_fork` tracepoint and the `task_alloc`
+ * LSM hook. When using only one of them, some child processes created inside
+ * containers were missing. To be sure we track everything, use both kind of
+ * programs for now. They both use the common handle_new_process() function.
+ */
+
+/*
+ * sched_process_fork - tracepoint program triggered by fork() function.
+ */
+SEC("tp_btf/sched_process_fork")
+int sched_process_fork(struct bpf_raw_tracepoint_args *args)
+{
+	struct task_struct *parent = (struct task_struct *)args->args[0];
+	struct task_struct *child = (struct task_struct *)args->args[1];
+
+	return handle_new_process(parent, child);
+}
+
+/*
+ * clone_audit - LSM program triggered by clone().
+ */
+SEC("lsm/task_alloc")
+int BPF_PROG(clone_audit, struct task_struct *task,
+	     unsigned long clone_flags, int ret_prev)
+{
+	struct task_struct *parent = BPF_CORE_READ(task, real_parent);
```

Comment:
Could this return `NULL`? If so, should we perform a NULL check on `parent`?

Reply:
I'm almost sure this should be never `NULL`, but doesn't hurt to check and log some error just in case. I will do that.

Reply:
Done